Step 1:
Presume there is a file named "US_births_1994-2003_CDC_NCHS.csv" in the current directory.  This file has the following columns:

year: Year (1994 to 2003).
month: Month (1 to 12).
date_of_month: Day number of the month (1 to 31).
day_of_week: Day of week (1 to 7).
births: Number of births that day.

Read the CSV file into a string.  Split the string on the newline character.  Display the first 10 values in the resulting list.

In [101]:
us_births = open("US_births_1994-2003_CDC_NCHS.csv").read().split("\n")
us_births[0:10]

['year,month,date_of_month,day_of_week,births',
 '1994,1,1,6,8096',
 '1994,1,2,7,7772',
 '1994,1,3,1,10142',
 '1994,1,4,2,11248',
 '1994,1,5,3,11053',
 '1994,1,6,4,11406',
 '1994,1,7,5,11251',
 '1994,1,8,6,8653',
 '1994,1,9,7,7910']

Step 2:
Create a function named read_csv() that:
Takes a single, required argument, a string representing the file name of the CSV file.
Reads the file into a string, splits the string on the newline character ("\n"), and removes the header row. Assign this list to string_list and create an empty list named final_list.
Uses a for loop to:
Iterate over string_list,
Create an empty list named int_fields,
Splits each row on the comma delimiter (,) and assigns the resulting list to string_fields,
Converts each value in string_fields to an integer and appends to int_fields,
Appends int_fields to final_list.
Returns final_list.

Use the read_csv() function to read in the file "US_births_1994-2003_CDC_NCHS.csv" and assign the result to cdc_list.

Display the first 10 rows of cdc_list to confirm it's a list of lists, containing only integer values, and no header row.

In [102]:
def read_csv(csv_file):
    string_list = open(csv_file).read().split("\n")
    string_list = string_list[1:]
    final_list = []
    
    for sl in string_list:
        int_fields = []
        string_fields = sl.split(",")
        for value in string_fields:
            int_fields.append(int(value))
    
        final_list.append(int_fields)
    
    return final_list

cdc_list = read_csv("US_births_1994-2003_CDC_NCHS.csv")
cdc_list[0:10]
                     


[[1994, 1, 1, 6, 8096],
 [1994, 1, 2, 7, 7772],
 [1994, 1, 3, 1, 10142],
 [1994, 1, 4, 2, 11248],
 [1994, 1, 5, 3, 11053],
 [1994, 1, 6, 4, 11406],
 [1994, 1, 7, 5, 11251],
 [1994, 1, 8, 6, 8653],
 [1994, 1, 9, 7, 7910],
 [1994, 1, 10, 1, 10498]]

Step 3: Create a function named month_births() that:
Takes a single, required argument, a list of lists.

Creates an empty dictionary, births_per_month, to store the monthly totals.

Uses a for loop to:
Iterate over the list of lists,
Extract the value in the month and births columns,
If the month value already exists as a key in births_per_month, the births value is added to the existing value,
If the month value doesn't exist as a key in births_per_month, it's created and the associated value is the births value.
After the loop, return the births_per_month dictionary.

Use the month_births() function to calculate the monthly totals for the dataset and assign the result to cdc_month_births. Display the dictionary.

In [103]:
def month_births(data):
    births_per_month = {}
    
    for row in data:
        month = row[1] #second column in list of lists
        births = row[4] #fifth column in list of lists
        
        if month in births_per_month:
            births_per_month[month] = births_per_month[month] + 1
        else:
            births_per_month[month] = births
    return births_per_month

cdc_month_births = month_births(cdc_list)
cdc_month_births

        

{1: 8405,
 2: 12036,
 3: 12436,
 4: 10929,
 5: 8454,
 6: 12648,
 7: 12763,
 8: 11737,
 9: 12482,
 10: 9660,
 11: 12615,
 12: 11990}

Step 4: Create a function named dow_births() that takes a single, required argument (a list of lists) and returns a dictionary containing the total number of births for each unique value of the day_of_week column.

Use the dow_births() function to return the day-of-week totals for the dataset and assign the result to cdc_day_births. Display the dictionary.

In [104]:
def dow_births(data):
     births_per_dow = {}
    
     for row in data:
        dow = row[3] #fourth column in list of lists
        births = row[4] #fifth column in list of lists
        
        if dow in births_per_dow:
            births_per_dow[dow] = births_per_dow[dow] + 1
        else:
            births_per_dow[dow] = births
    
     return births_per_dow

cdc_day_births = dow_births(cdc_list)
cdc_day_births

{1: 10663, 2: 11769, 3: 11574, 4: 11926, 5: 11771, 6: 8617, 7: 8293}

Step 5: Create a function named calc_counts() that:
Takes two, required parameters:
data: a list of lists
column: the column number we want to calculate the totals for
Populates and returns a dictionary containing the total number of births for each unique value in the column at position column.

Use the calc_counts() function to:
Return the yearly totals for the dataset and assign the result to cdc_year_births.
Return the monthly totals for the dataset and assign the result to cdc_month_births.
Return the day-of-month totals for the dataset and assign the result to cdc_dom_births.
Return the day-of-week totals for the dataset and assign the result to cdc_dow_births.

In [105]:
#create a function that works for all columns in dataset
def calc_counts(data, column):
    #populates and returns a dictionary containing the total number
    # of births for each unique value in the column at position <column>
    sums_dict = {}
    
    for row in data:
        col_value = row[column]
        births = row[4]
        
        if col_value in sums_dict:
            sums_dict[col_value] = sums_dict[col_value] + births
        else:
            sums_dict[col_value] = births
            
    return sums_dict

#return totals below for each column in dataset
cdc_year_births = calc_counts(cdc_list, 0)
cdc_month_births = calc_counts(cdc_list, 1)
cdc_dom_births = calc_counts(cdc_list, 2)
cdc_dow_births = calc_counts(cdc_list, 3)

In [106]:
#display value calculated for births by year
cdc_year_births

{1994: 3952767,
 1995: 3899589,
 1996: 3891494,
 1997: 3880894,
 1998: 3941553,
 1999: 3959417,
 2000: 4058814,
 2001: 4025933,
 2002: 4021726,
 2003: 4089950}

In [107]:
#dispaly value calculated for births by month
cdc_month_births

{1: 3232517,
 2: 3018140,
 3: 3322069,
 4: 3185314,
 5: 3350907,
 6: 3296530,
 7: 3498783,
 8: 3525858,
 9: 3439698,
 10: 3378814,
 11: 3171647,
 12: 3301860}

In [108]:
#display value calculated for births by day of month
cdc_dom_births

{1: 1276557,
 2: 1288739,
 3: 1304499,
 4: 1288154,
 5: 1299953,
 6: 1304474,
 7: 1310459,
 8: 1312297,
 9: 1303292,
 10: 1320764,
 11: 1314361,
 12: 1318437,
 13: 1277684,
 14: 1320153,
 15: 1319171,
 16: 1315192,
 17: 1324953,
 18: 1326855,
 19: 1318727,
 20: 1324821,
 21: 1322897,
 22: 1317381,
 23: 1293290,
 24: 1288083,
 25: 1272116,
 26: 1284796,
 27: 1294395,
 28: 1307685,
 29: 1223161,
 30: 1202095,
 31: 746696}

In [109]:
#display value calculated for births by date of week
cdc_dow_births

{1: 5789166,
 2: 6446196,
 3: 6322855,
 4: 6288429,
 5: 6233657,
 6: 4562111,
 7: 4079723}

Step 6: Here are some suggestions for next steps:

Write a function that can calculate the min and max values for any dictionary that's passed in.

Write a function that extracts the same values across years and calculates the differences between consecutive values to show if number of births is increasing or decreasing.
For example, how did the number of births on Saturday change each year between 1994 and 2003?

Find a way to combine the CDC data with the SSA data, which you can find here. Specifically, brainstorm ways to deal with the overlapping time periods in the datasets.

In [110]:
# function to extract the same values across years;
# also calculate the differences between consecutive values
# to show if number of births is increasing or decreasing

#let's see how the number of Monday births changed between 1994
#and 2003

monday_births = {}

for row in cdc_list:
    year = row[0]
    day_of_week = row[3]
    births = row[4]
    
    if day_of_week == 2: #monday
        if year in monday_births:
            monday_births[year] += 1
        else:
            monday_births[year] = births
        
monday_births        

{1994: 11299,
 1995: 10050,
 1996: 9460,
 1997: 11666,
 1998: 11894,
 1999: 12055,
 2000: 12754,
 2001: 10397,
 2002: 7842,
 2003: 12449}

Conclusion:

Monday births do not show a lot of deviation between the years 1994 and 2003.  The least number of Monday births in this range of years was 2002 with 7,842 births.  The highest number of Monday births in this range of years was 12,754 in 2000.   Also of note is that the number of Monday births hovered around 12,000 between 1997 and 2000. 